azurelinuxagent/common/protocol/wire.py

```diff
@@ -774,36 +771,45 @@ def update_host_plugin_from_goal_state(self):
         """
         Fetches a new goal state and updates the Container ID and Role Config Name of the host plugin client
         """
-        goal_state = GoalState.fetch_goal_state(self)
+        goal_state = GoalState(self)
         self._update_host_plugin(goal_state.container_id, goal_state.role_config_name)
 
-    def update_goal_state(self, forced=False):
+    def update_goal_state(self, force_update=False):
         """
-        Updates the goal state if the incarnation changed or if 'forced' is True
+        Updates the goal state if the incarnation or etag changed or if 'force_update' is True
         """
         try:
-            if self._goal_state is None or forced:
-                new_goal_state = GoalState.fetch_full_goal_state(self)
-            else:
-                new_goal_state = GoalState.fetch_full_goal_state_if_incarnation_different_than(self, self._goal_state.incarnation)
+            updated = False
+
+            goal_state = GoalState(self)
+
+            # Always update the hostgaplugin, since the agent may issue requests to it even if there are no other changes
+            # in the goal state (e.g. report status)
+            self._update_host_plugin(goal_state.container_id, goal_state.role_config_name)
+
+            if force_update:
+                logger.info("Forcing an update of the goal state..")
 
-            if new_goal_state is not None:
-                self._goal_state = new_goal_state
-                self._update_host_plugin(new_goal_state.container_id, new_goal_state.role_config_name)
-                if conf.get_enable_fast_track():
-                    self.update_extensions_goal_state()
+            if force_update or self._goal_state is None or self._goal_state.incarnation != goal_state.incarnation:
+                goal_state.fetch_full_goal_state(self)
+                self._goal_state = goal_state
+                updated = True
+
+            if conf.get_enable_fast_track():
+                if self._update_extensions_goal_state(force_update):
+                    updated = True
+
+            if updated:
                 self._save_goal_state()
 
         except Exception as exception:
             raise ProtocolError("Error fetching goal state: {0}".format(ustr(exception)))
 
-    def update_extensions_goal_state(self):
+    def _update_extensions_goal_state(self, force_update):
```

Comment:
Sometimes I read this as "update_extension_config" and get confused. Is there a more explicit name that we can give this? Something like "fetch_extension_config_via_fast_track" or something?

Reply:
Yup, this should become clearer on the next (or next+1) PR. In terms of queries, we need to get the vmSettings from hostgaplugin and/or the ExtensionsConfig from wireserver. We need a common representation for both, that will be named ExtensionsGoalState, which will be updated by update_extensions_config, which underneath will invoke fetch_vm_settings and fetch_extensions_config.